clustermesh: add service export read path in clustermesh/operator

MrFreezeex opened the pull request:
This commit adds the read path support for service exports inside the clustermesh/operator package like there is currently for regular services. We also copy a strip down/adapted version of GlobalServiceCache for service exports called GlobalServiceExportCache.

These new interfaces will be used for the MCS-API service import sync feature.

Related to #32712 #27902
Follow up to #32972

giorio94 commented on pkg/clustermesh/operator/remote_cluster.go:
```diff
@@ -62,6 +65,15 @@ func (rc *remoteCluster) Run(ctx context.Context, backend kvstore.BackendOperati
+			rc.remoteServiceExports.Watch(ctx, backend, path.Join(adapter(mcsapitypes.ServiceExportStorePrefix), rc.name))
+		})
+	} else {
+		// Drain the remote service exports in case the remote cluster no longer supports them
```
Should ServiceExports be marked synced in this case as well, to unblock the potential caller of `ServiceExportsSynced`?

giorio94 commented on pkg/clustermesh/operator/remote_cluster_test.go:
```diff
@@ -0,0 +1,177 @@
+	timeout = 5 * time.Second
+)
+
+func TestRemoteClusterStatus(t *testing.T) {
```
It would be great adding one more test to check that the different hooks are called correctly when triggering the corresponding action.

giorio94 commented on pkg/clustermesh/operator/service_exports.go:
```diff
@@ -0,0 +1,214 @@
+	scopedLog := r.log.WithFields(logrus.Fields{logfields.ServiceExportName: svcExport.String()})
+	scopedLog.Debug("Received remote service export delete event")
+
+	// Short-circuit the deletion logic if the service export was not present
```
Nit. Short-circuiting should not be necessary in this case, as update events are not filtered.

giorio94 commented on pkg/clustermesh/operator/service_exports.go:
```diff
@@ -0,0 +1,214 @@
+	clusterSvcExports := []*mcsapitypes.MCSAPIServiceSpec{}
+
+	if globalSvcExports, ok := c.byNamespace[namespace]; ok {
+		for globalSvcExport := range globalSvcExports {
```
```suggestion
		for globalSvcExport := range c.byNamespace[namespace] {
```

Nit. Dropping the surrounding if, for simplicity.

giorio94 commented on pkg/clustermesh/operator/service_exports.go:
```diff
@@ -0,0 +1,214 @@
+	if !ok {
+		globalSvcExport = newGlobalServiceExport()
+		c.byName[svcExport.NamespacedName()] = globalSvcExport
+		scopedLog.Debugf("Created global service export %s", svcExport.NamespacedName())
```
Nit. I'd personally drop all the logs from `GlobalServiceExportCache`, as mostly redundant with respect to the ones already output by the observer (I know the service cache emits them as well, and they don't look great there either IMO).

giorio94 commented on pkg/clustermesh/operator/service_exports.go:
```diff
@@ -0,0 +1,214 @@
+}
+
+// must be called with c.mutex held
+func (c *GlobalServiceExportCache) delete(globalServiceExport *GlobalServiceExport, clusterName string, serviceExportNN types.NamespacedName) bool {
```
Nit. I'd probably merge this with `OnDelete`, to make it symmetric of `OnUpdate`.

giorio94 commented on pkg/clustermesh/operator/service_exports.go:
```diff
@@ -0,0 +1,214 @@
+	}
+}
+
+type GlobalServiceExportCache struct {
```
I wonder whether a structure like:

```go
type ServiceExportsByNamespace map[string]ServiceExportsByName
type ServiceExportsByName      map[string]ServiceExportsByCluster
type ServiceExportsByCluster   map[string]*mcsapitypes.MCSAPIServiceSpec

type GlobalServiceExportCache struct {
    ...
    cache ServiceExportsByNamespace
    ...
}
```

could be a bit easier to understand. WDYT?

MrFreezeex replied:
Yep that sounds clearer indeed, thanks! I just add to track the size separately to not iterate over everything every time we want to track the size (i.e.: metrics) and that incrementing float may go haywire but apart from that the code is much easier I think :D

giorio94 commented on pkg/clustermesh/operator/service_exports.go:
It would be great if you could add a couple of unit tests to cover the different `GlobalServiceExportCache` operations, given that there are a few subtleties in the corner cases.

giorio94 reviewed (commented):
Thanks! The logic looks good to me, just a couple of nits inline, and the suggestion for adding a few more unit/integration tests.

> I went with the approach of duplicating the global service cache into a totally different struct for service export since what we want to expose is slightly different (primarily get all the service exports in a namespace to trigger some reconciliation if a new namespace is created and get all the service exports with a specific name/namespace to actually perform the service import reconciliation) and the struct that we want to share (service vs service export) is slightly different (no check if the service is shared for instance). Let me know if you have any general concern with that approach!

Overall sounds reasonable to me, and I agree that the structure is a bit different. Another possibility could be generalizing a bit more the cache using generics, and attempting to reuse it in both cases. Maybe if the `cache[namespace][name][cluster]` approach works... I'd instead keep the observer separate, given that as you also mentioned the logic is pretty different (no filtering).

> Also I am not entirely sure if this needs to be splitted into multiple commits as it's mainly one feature in one go package... Happy to split this into multiple commit it if you have any suggestions on that matter (or anything else actually :D).

I'd probably suggest to have one commit introducing/modifying the global cache, and another with all the other modifications, just for a bit more separation.

MrFreezeex committed:
clustermesh: add GlobalServiceExportCache

This commit adds a strip down/adapted version of GlobalServiceCache for
service exports called GlobalServiceExportCache which is going to be used
for reading service exports from a controller.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
clustermesh: add service export read path in clustermesh/operator

This commit adds the read path support for service exports inside the
clustermesh/operator package like there is currently for regular
services. This is going to be used soon by the clustermesh/mcsapi package for
the ServiceImport sync support.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

giorio94 commented on pkg/clustermesh/operator/remote_cluster.go:
```diff
@@ -62,6 +65,19 @@ func (rc *remoteCluster) Run(ctx context.Context, backend kvstore.BackendOperati
+		// Also mimic that service exports are synced so that consumer may
+		// use remote service exports even if service exports are disabled or
+		// not supported by the remote cluster
+		rc.remoteServiceExports.Synced()
```
Hmm, this doesn't look correct, being a getter. You probably wanted to call `rc.synced.serviceExports.Stop()`.

MrFreezeex replied:
Ahh yes not the right method indeed! :sweat_smile:

giorio94 commented on pkg/clustermesh/operator/remote_cluster_test.go:
```diff
@@ -0,0 +1,235 @@
+		),
+	}
+
+	clusterAddCalledCount := 0
```
I'd use atomics here to prevent possible race conditions.

giorio94 commented on pkg/clustermesh/operator/service_exports.go:
```diff
@@ -0,0 +1,166 @@
+	for name := range c.cache[namespace] {
+		clusterSvcExportNames = append(clusterSvcExportNames, name)
+	}
+	return clusterSvcExportNames
```
Nit. return maps.Keys(c.cache[namespace])

giorio94 commented on pkg/clustermesh/operator/service_exports.go:
```diff
@@ -0,0 +1,166 @@
+	if len(c.cache[svcExport.Namespace]) != 0 {
+		return true
+	}
+	c.cache[svcExport.Namespace] = nil
```
Nit. Looks superfluous

MrFreezeex replied:
Woops indeed!

giorio94 commented on pkg/clustermesh/operator/service_exports_test.go:
```diff
@@ -0,0 +1,105 @@
+		Namespace: "default",
+	})
+	require.EqualValues(t, 1, globalServiceExports.Size())
+	require.EqualValues(t, []string{"reset"}, globalServiceExports.GetServiceExportsName("default"))
```
Nit. Could additionally assert that looking up a non existing namespace returns no elements.

giorio94 commented on pkg/clustermesh/operator/service_exports.go:
```diff
@@ -0,0 +1,166 @@
+	_, ok = svcExportsByCluster[svcExport.Cluster]
+	if !ok {
+		c.size += 1
+		c.metricTotalGlobalServiceExports.Set(float64(c.size))
```
IIRC, the same metric for global services counts the number of namespace/name pairs, not multiplied by the number of clusters. You may want to match that here as well.

MrFreezeex replied:
Ohhh yes good catch!

giorio94 reviewed (approved):
🚀 Just a few final nits inline

tommyp1ckles reviewed (approved):
looks good for cli
